src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+      for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+      {
+         Vst::ParameterInfo parameterInfo;
+         if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+         {
+            if(parameterInfo.flags & Vst::ParameterInfo::kIsReadOnly)
+               continue;
+
+            mComponentHandler->performEdit(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+            mEditController->setParamNormalized(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+         }
+      }
+   }
+   return true;
+}
+
+namespace
+{
+   unsigned CountChannels(Steinberg::Vst::IComponent* component,
+      const Steinberg::Vst::MediaTypes mediaType, 
+      const Steinberg::Vst::BusDirection busDirection,
+      const Steinberg::Vst::BusType busType)
+   {
+      using namespace Steinberg;
+
+      unsigned channelsCount{0};
+
+      const auto busCount = component->getBusCount(mediaType, busDirection);
+      for(auto i = 0; i < busCount; ++i)
+      {
+         Vst::BusInfo busInfo;
+         if(component->getBusInfo(mediaType, busDirection, i, busInfo) == kResultOk)
+         {
+            if(busInfo.busType == busType)
+               channelsCount += busInfo.channelCount;
+         }
+      }
+      return channelsCount;
+   }
+}
+
+unsigned VST3Effect::GetAudioInCount()
+{
+   return CountChannels(
+      mEffectComponent,
+      Steinberg::Vst::kAudio,
+      Steinberg::Vst::kInput,
+      Steinberg::Vst::kMain);
+}
+
+unsigned VST3Effect::GetAudioOutCount()
+{
+   return CountChannels(
+      mEffectComponent,
+      Steinberg::Vst::kAudio,
+      Steinberg::Vst::kOutput,
+      Steinberg::Vst::kMain);
+}
+
+int VST3Effect::GetMidiInCount()
+{
+   //Dummy
+   return 0;
+}
+
+int VST3Effect::GetMidiOutCount()
+{
+   //Dummy
+   return 0;
+}
+
+void VST3Effect::SetSampleRate(double rate)
+{
+   mSetup.sampleRate = rate;
+}
+
+size_t VST3Effect::SetBlockSize(size_t maxBlockSize)
+{
+   //Dummy
+   mSetup.maxSamplesPerBlock = static_cast<Steinberg::int32>(maxBlockSize);
+   return mSetup.maxSamplesPerBlock;
+}
+
+size_t VST3Effect::GetBlockSize() const
+{
+   return mSetup.maxSamplesPerBlock;
+}
+
+sampleCount VST3Effect::GetLatency()
+{
+   using namespace Steinberg;
+
+   auto latency { 0u };
+   
+   for(auto& realtimeProcessor : mRealtimeGroupProcessors)
+   {
+      Vst::IAudioProcessor* audioProcessor;
+      if(realtimeProcessor->mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, (void**)&audioProcessor) == kResultOk)
+         latency = std::max(
+            latency,
+            static_cast<decltype(latency)>(audioProcessor->getLatencySamples())
+         );
+   }
+
+   return latency;
+}
+
+size_t VST3Effect::GetTailSize()
+{
+   using namespace Steinberg;
+   Vst::IAudioProcessor* audioProcessor;
+   if(mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, reinterpret_cast<void**>(&audioProcessor)) == kResultOk)
+      return audioProcessor->getTailSamples();
```

Comment:
Why does GetLatency() take a maximum over the group of processors but this function does not?

The GetTailSize() function isn't yet used, and there are some old comments from Leland about that, but I'm not sure what they should be used for, unless maybe they matter in generation of sounds.

Reply:
Well, probably neither of them doesn't make sense. Especially GetTailSize(). As I understand tail is used to process effects that produce more samples than they consume (like reverb or delay). And the problem is that different tracks may have different sample rates thus, different tail sizes. Block size, however, is always the same for each track/processor.